### PR TITLE
[feature](Cloud) Add azure's support to BE

### PR DESCRIPTION
**Pull request by ByteYue**
## Proposed changes

As one subsequent pr of #35307, this pr will link azure into BE, and implements the corresponding interface of ObjStorageClient for Azure.

### CODE REVIEW

**Review by gavinchou — APPROVED** [review]

File: be/test/io/fs/azure_test.cpp [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+    std::string accountName = config::test_s3_ak;
+    std::string accountKey = config::test_s3_sk;
+
+    auto cred =
+            std::make_shared<Azure::Storage::StorageSharedKeyCredential>(accountName, accountKey);
+
+    const std::string containerName = config::test_s3_bucket;
+    const std::string blobName = "sample-blob";
+    const std::string blobContent = "Fuck Azure!";
```
Comment (gavinchou) [inline be/test/io/fs/azure_test.cpp:64]: Remove the F word

File: be/src/io/fs/azure_obj_storage_client.cpp
Comment (gavinchou) [inline be/src/io/fs/azure_obj_storage_client.cpp:1]: How do we test each function implementation?

**Review by gavinchou — CHANGES_REQUESTED** [review]

File: be/src/io/fs/azure_obj_storage_client.cpp
Comment (ByteYue) [reply be/src/io/fs/azure_obj_storage_client.cpp:1]: I'll add enough ut to test them right after i've done all the feature test like batch delete test in azure.

File: be/src/util/s3_util.cpp [hunk 2]
```diff
@@ -156,9 +159,9 @@ S3ClientFactory& S3ClientFactory::instance() {
     return ret;
 }
 
-std::shared_ptr<Aws::S3::S3Client> S3ClientFactory::create(const S3ClientConf& s3_conf) {
+std::shared_ptr<io::ObjStorageClient> S3ClientFactory::create(const S3ClientConf& s3_conf) {
```
[the remark this answers is not in the document]
Comment (ByteYue) [reply be/src/util/s3_util.cpp:162]: I think it could be one future work, since change the class name might be error-prone at developing time.

**Comment by ByteYue** [comment]
run buildall

**Comment by ByteYue** [comment]
run beut

**Comment by ByteYue** [comment]
run beut

**Comment by ByteYue** [comment]
run buildall

**Comment by ByteYue** [comment]
run buildall

**Comment by ByteYue** [comment]
run buildall

**Comment by ByteYue** [comment]
run buildall

**Review by dataroaring — DISMISSED** [review]
LGTM

**Review by gavinchou — APPROVED** [review]

File: gensrc/thrift/AgentService.thrift [hunk 3]
```diff
@@ -61,6 +61,15 @@ enum TTabletType {
     TABLET_TYPE_MEMORY = 1
 }
 
+enum TObjStorageType {
+    AWS = 0,
```
Comment (platoneko) [inline gensrc/thrift/AgentService.thrift:65]:
```suggestion
    UNKNOWN = 0,
```

File: be/src/util/s3_util.h [hunk 4]
```diff
@@ -116,9 +121,9 @@ struct S3ClientConf {
 
     std::string to_string() const {
         return fmt::format(
-                "(ak={}, token={}, endpoint={}, region={}, max_connections={}, "
+                "(ak={}, token={}, endpoint={}, region={}, bucket={}, max_connections={}, "
                 "request_timeout_ms={}, connect_timeout_ms={}, use_virtual_addressing={}",
-                ak, token, endpoint, region, max_connections, request_timeout_ms,
+                ak, token, endpoint, region, bucket, max_connections, request_timeout_ms,
                 connect_timeout_ms, use_virtual_addressing);
     }
```
Comment (platoneko) [inline be/src/util/s3_util.h:128]:
```c++
static S3Conf get_s3_conf(const ObjectStoreInfoPB&);
static S3Conf get_s3_conf(const TS3StorageParam&);
```
在函数实现里面处理 unknown provider，注意前向兼容

**Comment by ByteYue** [comment]
run buildall

**Comment by ByteYue** [comment]
run buildall

**Comment by ByteYue** [comment]
run buildall

**Comment by ByteYue** [comment]
run buildall

**Comment by ByteYue** [comment]
run buildall

**Review by gavinchou — APPROVED** [review]

**Review by dataroaring — DISMISSED** [review]
LGTM

**Comment by ByteYue** [comment]
run buildall

**Review by dataroaring — APPROVED** [review]
LGTM

**Pull request merged**